src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.DecisionDagRewriter.cs

```diff
@@ -137,13 +137,39 @@ public override BoundNode VisitCall(BoundCall node)
                     bool mightMutate =
                         // might be a call to a local function that assigns something
                         node.Method.MethodKind == MethodKind.LocalFunction ||
-                        // or perhaps we are passing a variable by ref and mutating it that way
-                        !node.ArgumentRefKindsOpt.IsDefault;
+                        // or perhaps we are passing a variable by ref and mutating it that way, e.g. `int.Parse(..., out x)`
+                        !node.ArgumentRefKindsOpt.IsDefault ||
```

Comment:
>!node.ArgumentRefKindsOpt.IsDefault [](start = 24, length = 35)

Could this be an Invoke for a delegate backed by a lambda that modifies a local? Please add a test for a scenario like that.

Reply:
Note that it is only an assignment to a pattern variable of the switch case that we need be worried about. A test here for a case like that could never be satisfied (i.e. it would be dead code), because the assignment inside the lambda would be caught in VisitAssignment.

Reply:
> Note that it is only an assignment to a pattern variable of the switch case that we need be worried about. A test here for a case like that could never be satisfied (i.e. it would be dead code), because the assignment inside the lambda would be caught in VisitAssignment.

It is hard to say anything without seen the code, let's have a test that invokes a delegate like that and we'll go from there.

---
In reply to: [405756878](https://github.com/dotnet/roslyn/pull/42972#discussion_r405756878) [](ancestors = 405756878)

Reply:
You say that its hard to say anything without seeing the code, but I don't know precisely what code it is you're wanting to see.  Are you asking for a delegate defined in a when clause that contains an assignment to the pattern variable, which is then invoked in that or a subsequent when clause?

---
In reply to: [405762952](https://github.com/dotnet/roslyn/pull/42972#discussion_r405762952) [](ancestors = 405762952,405756878)

Reply:
> Are you asking for a delegate defined in a when clause that contains an assignment to the pattern variable, which is then invoked in that or a subsequent when clause?

I am asking for a test that would get us here with an invocation of a delegate that modifies "an interesting" local. The delegate could be based on a lambda or a local function. Whatever is the "worst" situation.

---
In reply to: [405803545](https://github.com/dotnet/roslyn/pull/42972#discussion_r405803545) [](ancestors = 405803545,405762952,405756878)

Reply:
Could you please point me to the added test(s)

---
In reply to: [405811560](https://github.com/dotnet/roslyn/pull/42972#discussion_r405811560) [](ancestors = 405811560,405803545,405762952,405756878)

Reply:
It would also be good if you could explain why no changes are necessary here for correct behavior.

---
In reply to: [406984302](https://github.com/dotnet/roslyn/pull/42972#discussion_r406984302) [](ancestors = 406984302,405811560,405803545,405762952,405756878)

Reply:
Added test `DoNotShareTempMutatedThroughLambda_01`

There are two ways in which variables are shared.
- The input variable to a switch is shared with pattern-matching temps if there are no when clauses at all. Therefore we don't have to worry about lambdas mutating them in a when clause.
- The pattern variables declared in the switch sections can be shared with pattern-matching temps if there is no assignment in a when clause. We don't have to worry about a lambda mutating them because that lambda would have to appear in the when clause and contain an assignment.

---
In reply to: [406985341](https://github.com/dotnet/roslyn/pull/42972#discussion_r406985341) [](ancestors = 406985341,406984302,405811560,405803545,405762952,405756878)

Reply:
Activating pending addition of suggested tests.

---
In reply to: [407003133](https://github.com/dotnet/roslyn/pull/42972#discussion_r407003133) [](ancestors = 407003133,406985341,406984302,405811560,405803545,405762952,405756878)